File: internal/chart/heatmap.go
package chart

import (
	"encoding/xml"
	"fmt"
	"io"
	"math"

	"github.com/morteza-sakifard/market-data-lab/internal/core"
)

const (
	domWidth   = 80.0
	heatBidCol = "#2563eb"
	heatAskCol = "#dc2626"
	domLastCol = "#d97706"
)

// HeatmapView is resting size by bar index and grouped price. A nil
// or empty value leaves the step-8 candle layout unchanged.
// Columns is aligned with View.Bars; extra bars are skipped.
type HeatmapView struct {
	TicksPerRow int
	Columns     []HeatColumn
}

// HeatColumn is one time slot. Cells are already grouped and sorted.
type HeatColumn struct {
	Cells []HeatCell
}

// HeatCell is one grouped price. Bid/Ask are resting size, not trades.
type HeatCell struct {
	Price    core.Ticks
	Bid, Ask core.Qty
}

// DOMView is the live ladder. Levels are high price first.
type DOMView struct {
	Levels  []DOMLevel
	LastPx  core.Ticks
	LastQty core.Qty
}

// DOMLevel is one ladder row: bid size, ask size, last print at Price.
type DOMLevel struct {
	Price    core.Ticks
	Bid, Ask core.Qty
	Last     core.Qty
}

func hasHeatmap(v View) bool {
	if v.Heatmap == nil {
		return false
	}
	for _, col := range v.Heatmap.Columns {
		if len(col.Cells) > 0 {
			return true
		}
	}
	return false
}

func hasDOM(v View) bool {
	return v.DOM != nil && len(v.DOM.Levels) > 0
}

func writeHeatmap(w io.Writer, hm *HeatmapView, sc Scale) {
	if hm == nil {
		return
	}
	step := hm.TicksPerRow
	if step < 1 {
		step = 1
	}
	h := math.Abs(sc.Y(0) - sc.Y(core.Ticks(step)))
	if h < 1 {
		h = 1
	}
	var maxV core.Qty
	n := len(hm.Columns)
	if n > sc.N {
		n = sc.N
	}
	for i := 0; i < n; i++ {
		for _, c := range hm.Columns[i].Cells {
			if c.Bid > maxV {
				maxV = c.Bid
			}
			if c.Ask > maxV {
				maxV = c.Ask
			}
		}
	}
	if maxV <= 0 {
		return
	}
	half := sc.slotWidth() * 0.45
	if half < 1 {
		half = 1
	}
	for i := 0; i < n; i++ {
		x := sc.X(i)
		for _, c := range hm.Columns[i].Cells {
			y := sc.Y(c.Price) - h/2
			if c.Bid > 0 {
				writeHeatCell(w, x-half, y, half, h, heatBidCol, c.Bid, maxV, c.Price, "bid")
			}
			if c.Ask > 0 {
				writeHeatCell(w, x, y, half, h, heatAskCol, c.Ask, maxV, c.Price, "ask")
			}
		}
	}
}

func writeHeatCell(w io.Writer, x, y, width, height float64, color string, vol, max core.Qty, px core.Ticks, side string) {
	op := 0.12 + 0.70*float64(vol)/float64(max)
	fmt.Fprintf(w, "  <rect class=\"heat\" data-side=\"%s\" data-px=\"%d\" data-qty=\"%d\" x=\"%.1f\" y=\"%.1f\" width=\"%.1f\" height=\"%.1f\" fill=\"%s\" fill-opacity=\"%.2f\"/>\n",
		side, px, vol, x, y, width, height, color, op)
}

func writeDOM(w io.Writer, d *DOMView, inst core.Instrument, sc Scale, left, right float64) {
	if d == nil || len(d.Levels) == 0 {
		return
	}
	width := right - left
	if width < 1 {
		width = 1
	}
	tickH := math.Abs(sc.Y(0) - sc.Y(1))
	if tickH < 8 {
		tickH = 8
	}
	var maxV core.Qty
	for _, lv := range d.Levels {
		if lv.Bid > maxV {
			maxV = lv.Bid
		}
		if lv.Ask > maxV {
			maxV = lv.Ask
		}
	}
	mid := left + width*0.5
	barW := width * 0.22
	for _, lv := range d.Levels {
		y := sc.Y(lv.Price)
		if maxV > 0 {
			if lv.Bid > 0 {
				wbar := barW * float64(lv.Bid) / float64(maxV)
				if wbar < 1 {
					wbar = 1
				}
				fmt.Fprintf(w, "  <rect class=\"dom\" data-side=\"bid\" data-px=\"%d\" data-qty=\"%d\" x=\"%.1f\" y=\"%.1f\" width=\"%.1f\" height=\"%.1f\" fill=\"%s\" fill-opacity=\"0.45\"/>\n",
					lv.Price, lv.Bid, mid-4-wbar, y-tickH/2, wbar, tickH, heatBidCol)
			}
			if lv.Ask > 0 {
				wbar := barW * float64(lv.Ask) / float64(maxV)
				if wbar < 1 {
					wbar = 1
				}
				fmt.Fprintf(w, "  <rect class=\"dom\" data-side=\"ask\" data-px=\"%d\" data-qty=\"%d\" x=\"%.1f\" y=\"%.1f\" width=\"%.1f\" height=\"%.1f\" fill=\"%s\" fill-opacity=\"0.45\"/>\n",
					lv.Price, lv.Ask, mid+4, y-tickH/2, wbar, tickH, heatAskCol)
			}
		}
		if lv.Last > 0 {
			fmt.Fprintf(w, "  <circle class=\"dom\" data-side=\"last\" data-px=\"%d\" data-qty=\"%d\" cx=\"%.1f\" cy=\"%.1f\" r=\"2.5\" fill=\"%s\"/>\n",
				lv.Price, lv.Last, mid, y, domLastCol)
		}
		fmt.Fprintf(w, "  <text class=\"dom-px\" x=\"%.1f\" y=\"%.1f\" font-size=\"8\" fill=\"%s\" text-anchor=\"middle\">",
			mid, y+3, inkColor)
		_ = xml.EscapeText(w, []byte(inst.Text(lv.Price)))
		io.WriteString(w, "</text>\n")
	}
}

File: web/heatmap.js
// Heatmap and DOM — same colors and widths as internal/chart/heatmap.go.
(function () {
    const bid = "#2563eb";
    const ask = "#dc2626";
    const last = "#d97706";
    const ink = "#222222";
    const domW = 80;

    function maxSize(hm, i0, i1) {
        let max = 0;
        if (!hm || !hm.Columns) return 0;
        const lastCol = Math.min(i1, hm.Columns.length);
        for (let i = i0; i < lastCol; i++) {
            const cells = hm.Columns[i] && hm.Columns[i].Cells;
            if (!cells) continue;
            for (let j = 0; j < cells.length; j++) {
                if (cells[j].Bid > max) max = cells[j].Bid;
                if (cells[j].Ask > max) max = cells[j].Ask;
            }
        }
        return max;
    }

    function cellH(yOf, step) {
        return Math.max(1, Math.abs(yOf(0) - yOf(step)));
    }

    function drawColumn(ctx, box, vis, yOf, hm, index, max) {
        if (!hm || !hm.Columns || index < 0 || index >= hm.Columns.length) return;
        const col = hm.Columns[index];
        if (!col || !col.Cells || !col.Cells.length || max <= 0) return;
        const step = hm.TicksPerRow > 0 ? hm.TicksPerRow : 1;
        const h = cellH(yOf, step);
        const slot = vis.n > 0 ? (box.right - box.left) / vis.n : 0;
        const half = Math.max(1, slot * 0.45);
        const x = box.left + (index - vis.i0 + 0.5) * slot;
        col.Cells.forEach(function (c) {
            const y = yOf(c.Price) - h / 2;
            if (c.Bid > 0) {
                ctx.fillStyle = bid;
                ctx.globalAlpha = 0.12 + 0.70 * c.Bid / max;
                ctx.fillRect(x - half, y, half, h);
            }
            if (c.Ask > 0) {
                ctx.fillStyle = ask;
                ctx.globalAlpha = 0.12 + 0.70 * c.Ask / max;
                ctx.fillRect(x, y, half, h);
            }
        });
        ctx.globalAlpha = 1;
    }

    function draw(ctx, box, vis, yOf, hm) {
        const max = maxSize(hm, vis.i0, vis.i1);
        if (max <= 0) return;
        for (let i = vis.i0; i < vis.i1; i++) {
            drawColumn(ctx, box, vis, yOf, hm, i, max);
        }
    }

    function drawDOM(ctx, box, yOf, d) {
        if (!d || !d.Levels || !d.Levels.length || !box.domW) return;
        const left = box.domLeft;
        const width = box.domW;
        const mid = left + width * 0.5;
        let max = 0;
        d.Levels.forEach(function (lv) {
            if (lv.Bid > max) max = lv.Bid;
            if (lv.Ask > max) max = lv.Ask;
        });
        const tickH = Math.max(8, Math.abs(yOf(0) - yOf(1)));
        const barW = width * 0.22;
        ctx.font = "8px ui-monospace, monospace";
        ctx.textAlign = "center";
        d.Levels.forEach(function (lv) {
            const y = yOf(lv.Price);
            if (max > 0 && lv.Bid > 0) {
                let w = barW * lv.Bid / max;
                if (w < 1) w = 1;
                ctx.globalAlpha = 0.45;
                ctx.fillStyle = bid;
                ctx.fillRect(mid - 4 - w, y - tickH / 2, w, tickH);
            }
            if (max > 0 && lv.Ask > 0) {
                let w = barW * lv.Ask / max;
                if (w < 1) w = 1;
                ctx.globalAlpha = 0.45;
                ctx.fillStyle = ask;
                ctx.fillRect(mid + 4, y - tickH / 2, w, tickH);
            }
            ctx.globalAlpha = 1;
            if (lv.Last > 0) {
                ctx.fillStyle = last;
                ctx.beginPath();
                ctx.arc(mid, y, 2.5, 0, Math.PI * 2);
                ctx.fill();
            }
        });
        ctx.textAlign = "left";
    }

    function atHeat(hm, barI, price) {
        if (!hm || !hm.Columns || barI < 0 || barI >= hm.Columns.length) {
            return {bid: 0, ask: 0};
        }
        const cells = hm.Columns[barI].Cells || [];
        const step = hm.TicksPerRow > 0 ? hm.TicksPerRow : 1;
        const g = price - (price % step);
        for (let i = 0; i < cells.length; i++) {
            if (cells[i].Price === g) return {bid: cells[i].Bid, ask: cells[i].Ask};
        }
        return {bid: 0, ask: 0};
    }

    function atDOM(d, price) {
        if (!d || !d.Levels) return {bid: 0, ask: 0, last: 0};
        for (let i = 0; i < d.Levels.length; i++) {
            if (d.Levels[i].Price === price) {
                return {bid: d.Levels[i].Bid, ask: d.Levels[i].Ask, last: d.Levels[i].Last};
            }
        }
        return {bid: 0, ask: 0, last: 0, lastPx: d.LastPx, lastQty: d.LastQty};
    }

    window.MDL = window.MDL || {};
    window.MDL.heatmap = {
        width: domW,
        draw: draw,
        drawColumn: function (ctx, box, vis, yOf, hm, index) {
            drawColumn(ctx, box, vis, yOf, hm, index, maxSize(hm, vis.i0, vis.i1));
        },
        drawDOM: drawDOM,
        atHeat: atHeat,
        atDOM: atDOM,
    };
})();
